server/mqtt_ex_test.go

```diff
@@ -0,0 +1,421 @@
+			)
+		})
+	})
+}
+
+// makes a copy of bc
+func (bc mqttBenchContext) benchmarkPubSub(b *testing.B) {
+	// This test uses a single built-in topic, and a built-in publisher, so no
+	// reason to run it for topics and publishers.
+	m := mqttBenchDefaultMatrix.
+		NoTopics().
+		NoPublishers()
+
+	b.Run("PUBSUB", func(b *testing.B) {
+		m.runMatrix(b, bc, func(b *testing.B, bc *mqttBenchContext) {
+			bc.runCommand(b, "pubsub",
+				"--qos", strconv.Itoa(bc.QOS),
+				"--n", strconv.Itoa(b.N),
+				"--size", strconv.Itoa(bc.MessageSize),
+				"--num-subscribers", strconv.Itoa(bc.Subscribers),
+			)
+		})
+	})
+}
+
+// makes a copy of bc
+func (bc mqttBenchContext) benchmarkSubRet(b *testing.B) {
+	// This test uses a a built-in publisher, and it makes most sense to measure
+	// the retained message delivery "overhead" on a QoS0 subscription; without
+	// the extra time involved in actually subscribing.
+	m := mqttBenchDefaultMatrix.
+		NoPublishers().
+		QOS0Only()
+
+	b.Run("SUBRET", func(b *testing.B) {
+		m.runMatrix(b, bc, func(b *testing.B, bc *mqttBenchContext) {
+			bc.runCommand(b, "subret",
+				"--qos", strconv.Itoa(bc.QOS),
+				"--n", strconv.Itoa(b.N), // number of subscribe requests
+				"--num-subscribers", strconv.Itoa(bc.Subscribers),
+				"--num-topics", strconv.Itoa(bc.Topics),
+				"--size", strconv.Itoa(bc.MessageSize),
+			)
+		})
+	})
+}
+
+func mqttBenchLookupCommand(b *testing.B, name string) string {
+	b.Helper()
+	cmd, err := exec.LookPath(name)
+	if err != nil {
+		b.Skipf("%q command is not found in $PATH. Please `go install github.com/nats-io/meta-nats/apps/go/mqtt/...@latest` and try again.", name)
+	}
+	return cmd
+}
+
+func (bc mqttBenchContext) runCommand(b *testing.B, name string, extraArgs ...string) {
+	b.Helper()
+
+	args := append([]string{
+		name,
+		"-q",
+		"--servers", fmt.Sprintf("%s:%d", bc.Host, bc.Port),
+	}, extraArgs...)
+
+	cmd := exec.Command(bc.testCmdPath, args...)
+	stdout, err := cmd.StdoutPipe()
+	if err != nil {
+		b.Fatalf("Error executing %q: %v", cmd.String(), err)
+	}
+	defer stdout.Close()
+	errbuf := bytes.Buffer{}
+	cmd.Stderr = &errbuf
+	if err = cmd.Start(); err != nil {
+		b.Fatalf("Error executing %q: %v", cmd.String(), err)
+	}
+	r := &MQTTBenchmarkResult{}
+	if err = json.NewDecoder(stdout).Decode(r); err != nil {
+		b.Fatalf("failed to decode output of %q: %v\n\n%s", cmd.String(), err, errbuf.String())
+	}
+	if err = cmd.Wait(); err != nil {
+		b.Fatalf("Error executing %q: %v\n\n%s", cmd.String(), err, errbuf.String())
+	}
+
+	r.report(b)
+}
+
+func (bc mqttBenchContext) initServer(b *testing.B) {
+	b.Helper()
+	bc.runCommand(b, "pubsub",
+		"--id", "__init__",
+		"--qos", "0",
+		"--n", "1",
+		"--size", "100",
+		"--num-subscribers", "1")
+}
+
+func (bc *mqttBenchContext) startServer(b *testing.B, disableRMSCache bool) func() {
+	b.Helper()
+	b.StopTimer()
+	o := testMQTTDefaultOptions()
+	o.MQTT.disableRetainedMessageCache = disableRMSCache
```

Comment:
Would remove that of course, instead, at the top of the function add:
```
testDisableRMSCache = true
```
and in the defer function for the shutdown:
```
return func() {
    testMQTTShutdownServer(s)
    testDisableRMSCache = false
}
```

But, this would work only if the tests are running sequentially, not in parallel (at least from the same test suite process), because then it could cause races and behave unexpectedly).